Fix: support whitespace in data providers

carlfriedrich opened the pull request:
## 📚 Description

This PR is a potential fix for #478. It adds a solution for supporting whitespace in data provider values.

## 🔖 Changes

The PR adds a global `data_set` function for declaring data sets in data providers, superseding the previous `echo` solution with a more robust way preserving quotes.

It furthermore changes the way data from data providers is parsed.
Previously the values were simply read using `IFS=" "`, which split the
input on spaces. This made values containing spaces impossible.
Now values are parsed using a combination of `eval` and `set`, which
respects quotes in the input data and hence makes contained spaces
possible. We can also provide empty strings as input data now.
The new parsing is backwards-compatible, so declaring data values using `echo` still works like before.

## ✅ To-do list

- [ ] I updated the `CHANGELOG.md` to reflect the new feature or fix
- [ ] I updated the documentation to reflect the changes

carlfriedrich commented:
Hey maintainers. 👋 I am willing to update changelog and documentation, but first I would like to get a first review on my approach for this fix.

Specific questions I have:

1. Do you like the new global function `data_set` for declaring data values in data provider functions? Is the naming appropriate?
2. Is the new parsing mechanism using `eval` suitable or do you have any concerns about it?
   Generally speaking, `eval` bears the risk of command injection. However, when used in conjunction with `data_set`, all input tokens are quoted, so this should be safe IMO, as long as we update the documentation to consequently use `data_set` instead of `echo` in data providers.

I am happy to hear your feedback on this. If we can agree on this solution I will add commits to update changelog and documentation.

Chemaclass reviewed (approved):
Awesome work, thank you!

Chemaclass commented:
@carlfriedrich please **update the docs** and add one liner to the `CHANGELOG.md` before merging this PR

carlfriedrich commented:
@Chemaclass 

I added documentation and changelog. I also added a commit which updates all bashunit test cases to use the new `data_set` function in data providers. This led to one other failing test case, which I pushed a fix for. I am not sure why this particular test is implemented the way it is (why do we [use a subshell to find the total tests](https://github.com/TypedDevs/bashunit/blob/05d5894946db498fd60268a82ea0890a59693fbe/tests/unit/helpers_test.sh#L260)?), that's why figuring out what was going on here took quite some time.

However, from my side this is merge-ready now.

Chemaclass reviewed (approved):
nice docs!

carlfriedrich commented:
@Chemaclass Thanks a lot for your fast and valuable feedback! Was fun contributing to this project. Looking forward to using it even more. 😃 
Can you already estimate when this PR will land in a release?

Chemaclass commented:
I think I will create a new release with the changes we already have by this weekend or next weekend. Until then, you can already benefit from the latest changes installing the beta (which is installing the code from main branch). 
> Here is how: https://bashunit.typeddevs.com/installation#define-custom-tag-and-folder

carlfriedrich commented:
Ah, great to know! Thanks a lot again!
